app/codecs/CirceDecoders.scala

```diff
@@ -38,25 +36,26 @@ object CirceDecoders {
   implicit val decodeCountry: Decoder[Country] =
     Decoder.decodeString.emap { code => CountryGroup.countryByCode(code).toRight(s"Unrecognised country code '$code'") }
 
+  //Payment fields are input from support-frontend
+  implicit val payPalPaymentFieldsCodec: Codec[PayPalPaymentFields] = deriveCodec
+  implicit val stripePaymentFieldsCodec: Codec[StripePaymentFields] = deriveCodec
+  implicit val directDebitPaymentFieldsCodec: Codec[DirectDebitPaymentFields] = deriveCodec
+
+  //noinspection ConvertExpressionToSAM
   implicit val encodePaymentFields: Encoder[PaymentFields] = new Encoder[PaymentFields] {
-    private val stripeEncoder = deriveEncoder[StripePaymentFields]
-    private val paypalEncoder = deriveEncoder[PayPalPaymentFields]
-    override def apply(a: PaymentFields): Json = {
-      a.fold(stripeEncoder.apply, paypalEncoder.apply)
+    override final def apply(a: PaymentFields): Json = a match {
+      case p: PayPalPaymentFields => Encoder[PayPalPaymentFields].apply(p)
+      case s: StripePaymentFields => Encoder[StripePaymentFields].apply(s)
+      case d: DirectDebitPaymentFields => Encoder[DirectDebitPaymentFields].apply(d)
     }
   }
 
-  implicit val paymentFields: Decoder[PaymentFields] = {
-    val stripeFields = deriveDecoder[StripePaymentFields].map(_.asLeft[PayPalPaymentFields])
-    val payPalFields = deriveDecoder[PayPalPaymentFields].map(_.asRight[StripePaymentFields])
-    stripeFields or payPalFields
-  }
-
-  implicit val requestPaymentFields: Decoder[Either[StripePaymentToken, PayPalPaymentFields]] = {
-    val stripeFields = deriveDecoder[StripePaymentToken].map(_.asLeft[PayPalPaymentFields])
-    val payPalFields = deriveDecoder[PayPalPaymentFields].map(_.asRight[StripePaymentToken])
-    stripeFields or payPalFields
-  }
+  implicit val decodePaymentFields: Decoder[PaymentFields] =
```

Comment:
This might be slightly nicer (from https://stackoverflow.com/questions/42165460/how-to-decode-an-adt-with-circe-without-disambiguating-objects)

```
import cats.syntax.functor._


  implicit val decodePaymentFields: Decoder[PaymentFields] = {
    List[Decoder[PaymentFields]](
      Decoder[PayPalPaymentFields].widen,
      Decoder[StripePaymentFields].widen,
      Decoder[DirectDebitPaymentFields].widen
    ).reduce(_ or _)
  }
```

Reply:
Updated in 7e0e05a. Tested the UK flows are they keep working.